stable/database/files/nuosm

```diff
@@ -33,312 +39,499 @@ if [ $crashcount -ge 1 ]; then
   fi
 fi
 
-export NUOCMD DB_NAME NUODB_STORAGE_PASSWORDS_DIR
+export DB_NAME NUODB_STORAGE_PASSWORDS_DIR
 
 #=======================================
 # function - report an error and exit
 #
 function die() {
-  # cleanup
-  $NUOCMD set value --key $first_req --value '' --expected-value $HOSTNAME
-
   retval=$1
   shift
-  echo "$@"
+  error="$@"
+  [ -n "$wotIdid" ] && error="Error while ${wotIdid}: ${error}"
+  log "$error"
+
+  # cleanup
+  nuocmd set value --key $first_req --value '' --expected-value $HOSTNAME
+  nuocmd set value --key $startup_key/$DB_NAME --value '' --expected-value $HOSTNAME
+
+  # release my archive if it is locked
+  releaseLockedArchive
+
   exit $retval
 }
 
 #=======================================
-# function - perform archive restore
+# function - log messages
+#
+function log() {
+  local err="$@"
+  echo "$err" | tee -a $LOGFILE
+}
+
+#=======================================
+# function - trace workflow
+#
+function trace() {
+  wotIdid="$@"
+  [ -n "$1" -a -n "$NUODB_DEBUG" ] && log "trace: $wotIdid"
+}
+
+#=======================================
+# function - wrap a log file around so it doesn't grow infinitely
+#
+function wrapLogfile() {
+  logsize=$( du -sb $LOGFILE | grep -o '^ *[0-9]\+' )
+  maxlog=5000000
+  log "logsize=$logsize; maxlog=$maxlog"
+  if [ ${logsize:=0} -gt $maxlog ]; then
+    lines=$(wc -l $LOGFILE)
+    tail -n $(( lines / 2 )) $LOGFILE > ${LOGFILE}-new
+    rm -rf $LOGFILE
+    mv $LOGFILE-new $LOGFILE
+    log "(nuosm) log file wrapped around"
+  fi
+}
+
+function isUrl() {
+  if echo $1 | grep -q '^[a-z]\+:/[^ ]\+'; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+function isRestoreSourceAvailable() {
+  if isUrl "$restore_source" || [ -d $NUODB_BACKUPDIR/$restore_source ]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+function releaseLockedArchive() {
+  trace "releasing my locked archive metadata"
+  locked_archive=$( nuocmd show archives --db-name $DB_NAME --removed --removed-archive-format "archive-id: {id}" | sed -En "/^archive-id: / {N; /$HOSTNAME/ s/^archive-id: ([0-9]+).*$/\1/; T; p}" | head -n 1 )
```

Comment:
I'm not familiar with the sed syntax used here, but this seems to assume that there is only one archive on the current host and that it has been started at least once. Can you include a comment describing what this doozy of a one-liner does (maybe @NikTJ777 can explain)?

Reply:
This is my understanding as well. It expects the `archiveId` on the first line followed by the exited process on the next line. AFAIK this is the case always in docker deployments as the process is started right after the archive is created by `nuodocker start sm`.
I will add a comment.